docs/design/2020-05-08-standardize-error-codes-and-messages.md

```diff
@@ -0,0 +1,204 @@
+
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+```
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- The markdown content can be previewed on writing.
+
+Tradeoff:
+- It's hard to implement (see example below)
+- It's not self-consistent (see example below)
+
+Tradeoff Example:
+
+```md
+## Code: 8005
+### Error
+Write Conflict, txnStartTS is stale
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+```
+
+As the syntax above, the 9005 block is the message part of 8005 block, so we expect it's result is the same as this toml:
+
+```toml
+[error.8005]
+error = '''Write Conflict, txnStartTS is stale'''
+message = '''Transactions in TiDB encounter write conflicts.'''
+workaround = '''
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+'''
+```
+
+In this case, we must define the grammar and write a parser. Writing parser increase much complexity. What's worse
+is that I found no grammar that fits it.
+
+#### Summary
+
+Through the above discussion, we recommend the toml version of metafile. 
+
+In addition, the error codes should be append only in case of conflict between versions.
+
+### The Error Definition
+
+In the discussion above, an error has at least 4 parts:
+- The error code: it's the identity of an error
+- The error field: it's the error itself the user can view in TiDB system
+- the message field: the description of the error, what happened and why happend?
+- the workaround filed: how to workaround this error
+
+Besides, we can append a optional tags field to it:
+```toml
+[error.9005]
+error = ""
+message = ""
+workaround = ""
+tags = ["tikv"]
+```
+
+The tags is used to classify errors.
+
+#### The Error Code Range
+
+For compatibility with MySQL protocol, the code transmitted through the mysql protocol should be number only, others can be a number with a prefix string.
+
+The code of each components looks like:
+- TiDB: [0, 9000), DB\d+
+- TiKV: [9000, 9010), KV\d+
+- PD: [9000, 9010), PD\d+
+- DM: DM\d+, DM-S-\d+
```

Comment:
@kennytm what about br & cdc?

Reply:
@lonng please add error code range for `mysql replication`

Reply:
BR, Lightning, Dumpling, and likely DM and even Binlog, CDC depend on several shared libraries. What component prefix should we associate with these?

* Table filter / black-white list
* External storage (S3, GCS r/w)
* Split & scatter 

If these two will have their own prefix, I suggest:

* `FILTER-[0-9]{2,}` / `EXTSTR-[0-9]{4,}` / `TOOL-[0-9]{4,}` / ??? - the common components
* `BR-[0-9]{4,}` - BR 
* `CDC-[0-9]{4,}` - CDC
* `LN-[0-9]{4,}` - Lightning (and Importer)
* `DP-[0-9]{4,}` - Dumpling

Otherwise, I suggest:

* `BR-[A-Z]-[0-9]{3,}` - BR 
* `CDC-[A-Z]-[0-9]{3,}` - CDC
* `LN-[A-Z]-[0-9]{3,}` - Lightning (and Importer)
* `DP-[A-Z]-[0-9]{3,}` - Dumpling

Reply:
I prefer the latter, let's not add too much TiDB specification to the base library of features